### PR TITLE
Make SocketPolicy a sealed interface

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: mockwebserver/src/main/kotlin/mockwebserver3/MockWebServer.kt [hunk 1]
```diff
@@ -939,10 +939,10 @@ class MockWebServer : Closeable {
       }
 
       when (socketPolicy) {
-        DISCONNECT_AT_END -> {
+        DisconnectAtEnd -> {
           stream.connection.shutdown(ErrorCode.NO_ERROR)
         }
-        DO_NOT_READ_REQUEST_BODY -> {
+        DoNotReadRequestBody -> {
```
Comment (yschimke) [inline mockwebserver/src/main/kotlin/mockwebserver3/MockWebServer.kt:945]: Seems a shame to not put things like http2ErrorCode into the DoNotReadRequestBody type?

File: mockwebserver/src/main/kotlin/mockwebserver3/SocketPolicy.kt [hunk 2]
```diff
@@ -49,70 +49,70 @@ enum class SocketPolicy {
...
-  DISCONNECT_DURING_REQUEST_BODY,
+  object DisconnectDuringRequestBody : SocketPolicy
 
   /** Close connection after writing half of the response body (if present). */
-  DISCONNECT_DURING_RESPONSE_BODY,
+  object DisconnectDuringResponseBody : SocketPolicy
 
   /**
    * Process the response without even attempting to reading the request body. For HTTP/2 this will
    * send [MockResponse.getHttp2ErrorCode] after the response body or trailers. For HTTP/1 this will
```
Comment (yschimke) [inline mockwebserver/src/main/kotlin/mockwebserver3/SocketPolicy.kt:80]: As above, how about making this a Data class?

**Review by yschimke — APPROVED** [review]
One API question.

**Comment by swankjesse** [comment]
You read my mind!
PR with data classes is coming up next.

**Comment by MinJunKweon** [comment]
Just curious, why did it change to sealed interface from enum class?

**Comment by swankjesse** [comment]
That allows us to model policy-specific parameters like `DoNotReadRequestBody.http2ErrorCode`.

**Pull request merged**